_posts/2020-07-16-building-drivers-on-zephyr.md

```diff
@@ -0,0 +1,589 @@
+
+Here's an example of a `Kconfig` from the `ncs/nrf/applications/asset_tracker` application
+
+```
+menu "Asset tracker"
+
+rsource "src/ui/Kconfig"
+
+config APPLICATION_WORKQUEUE_STACK_SIZE
+	int "Application workqueue stack size"
+	default 4096
+
+config APPLICATION_WORKQUEUE_PRIORITY
+	int "Application workqueue priority"
+	default SYSTEM_WORKQUEUE_PRIORITY
+
+menu "GPS"
+...
+```
+
+You can see that it references separate `Kconfig` files as a dependency. You can also see that some environment variables are set with default values. You can either add them to your project's `prj.conf` *or* edit them using `west build -t menuconfig`. If you're looking to make permanent changes, you should use the `prj.conf` method.
+
+The most important part about `Kconfig` files in regards to drivers? You can use these variables as flags for conditional compilation.
+
+As an example, here's what the `ncs/nrf/drivers/sensor/CMakeLists.txt` looks like:
+
+```cmake
+add_subdirectory_ifdef(CONFIG_BH1749 bh1749)
+add_subdirectory_ifdef(CONFIG_SENSOR_SIM sensor_sim)
+add_subdirectory_ifdef(CONFIG_PMW3360 pmw3360)
+add_subdirectory_ifdef(CONFIG_PAW3212 paw3212)
+```
+
+You may noice that all the variables begin with `CONFIG_`. All variables produced by `Kconfig` have this prefix. So, for example, `APPLICATION_WORKQUEUE_STACK_SIZE` becomes `CONFIG_APPLICATION_WORKQUEUE_STACK_SIZE.`
+
+In the `CMakeLists.txt` above, you can see that a subdirectory is only added if the driver is "enabled". `CONFIG_BH1749` happens to be the top-level environment variable to make that happen. You can also use this type of conditional compilation for your project code.
+
+### `zephyr/module.yml`
+
+`Kconfig` and `CMakeList.txt` are great but where do we start? How does `west` know where to start looking for them?
+
+The answer is a very unsuspecting file called `module.yml`. `module.yml` is usually found in the `/zephyr` subdirectory of a dependency. For example, here's the one for nRF Connect SDK located in `ncs/nrf/zephyr/`
+
+Example
+
+```
+build:
+  cmake: .
+  kconfig: Kconfig.nrf
+```
+
+The file tells `west` where to find the appropriate base `Kconfig` and `CMakeLists.txt`. Those are the starting point for further repository traversal. Without these, there's no way to use your driver code in your application!
+
+## Adding a drivers folder
+
+You now should understand `Kconfig`, `CMakeLists.txt`, and `module.yml`. We're ready for driver development!
+
+ In my case, since my driver work is ongoing, I opted to use my own `drivers` folder in my main project repository. You can always opt to start developing your drivers in `zephyr/drivers`. This is likely a better location especially if you want to release the driver for others to use. The downside is that it requires more work!
+
+In my case, here's what the directory structure looks like from my base repo:
+
+```terminal
+.
```

Comment:
Might call out the root directory location here too, just to be extra explicit
```suggestion
. # My Project Root, pyrinas
```

Reply:
Ahh good call. Fixed that!